Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50236 Apply clang-format to rpc cpp files
* #50029 Completely remove FutureMessage type
* #50028 Completely Remove FutureMessage from RPC agents
* #50027 Completely Remove FutureMessage from RPC cpp tests
* #50026 Remove FutureMessage from RPC request callback logic
* #50025 Completely remove FutureMessage from FaultyProcessGroupAgent
* #50024 Remove FutureMessage from sender TensorPipeAgent
* #50023 Remove FutureMessage from sender ProcessGroupAgent
* #50020 Completely remove FutureMessage from distributed autograd
* #50005 Completely remove FutureMessage from RPC TorchScript implementations
* #50004 Completely remove FutureMessage from RRef Implementations
* #49995 Replace FutureMessage with ivalue::Future in RpcAgent retry logic
* #49960 Replace FutureMessage with ivalue::Future in RRefContext
* **#49927 Replace FutureMessage with ivalue::Future in distributed/autograd/utils.***
* #49906 Let RpcAgent::send() return JitFuture

Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

mrshenli committed:
Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*"


Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*"


Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*"


Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*"


Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*"


Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*"


Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*"


Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -134,42 +134,40 @@ std::shared_ptr<FutureMessage> sendPythonRemoteCall(
-            jitFuture->markCompleted(
-                toIValue(futureResponseMessage->constValue()));
+            pyJitFuture->markCompleted(
+                toIValue(*messageJitFuture->value().toCustomClass<Message>()));
```
I'm not familiar with JIT intricacies, but to me this looks like we're casting the IValue to a Message instance, just to then convert it back to an IValue. What am I missing?

lw replied:
Another thing I'm not getting: the future we return has type `PyObject` but it still contains the same `Message` instance. Does this mean we're exposing Message to Python?

mrshenli replied:
The input IValue contains a message, and the output IValue contains a PyObject. The `toIValue` function deserializes the message based on the type and convert that into `py::object` and then wrap it into an IValue. Let me add more comments to clarify that.

lw replied:
I see, I guess I just assumed what `toIValue` did without bothering to go check.

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*"


Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*"


Differential Revision: [D25724241](https://our.internmc.facebook.com/intern/diff/D25724241)

[ghstack-poisoned]
